src/emulator/storage/upload.ts

```diff
@@ -0,0 +1,188 @@
+      objectId: request.objectId,
+      type: UploadType.MULTIPART,
+      path: this.stagingFileName(id, request.bucketId, request.objectId),
+      status: UploadStatus.FINISHED,
+      metadata: JSON.parse(request.metadataRaw),
+      size: sizeInBytes,
+      authorization: request.authorization,
+    };
+    this._uploads.set(upload.id, upload);
+
+    return upload;
+  }
+
+  /**
+   * Initializes a new ResumableUpload.
+   */
+  public startResumableUpload(request: StartResumableUploadRequest): Upload {
+    const id = uuidV4();
+    const upload: Upload = {
+      id: id,
+      bucketId: request.bucketId,
+      objectId: request.objectId,
+      type: UploadType.RESUMABLE,
+      path: this.stagingFileName(id, request.bucketId, request.objectId),
+      status: UploadStatus.ACTIVE,
+      metadata: JSON.parse(request.metadataRaw),
+      size: 0,
+      authorization: request.authorization,
+    };
+    this._uploads.set(upload.id, upload);
+    this._persistence.deleteFile(upload.path, /* failSilently = */ true);
+    return upload;
+  }
+
+  /**
+   * Appends bytes to an existing resumable upload.
+   * @throws {NotFoundError} if the resumable upload does not exist.
+   * @throws {NotActiveUploadError} if the resumable upload is not in the ACTIVE state.
+   */
+  public progressResumableUpload(uploadId: string, dataRaw: string): Upload {
+    const upload = this.findResumableUpload(uploadId);
+    if (upload.status !== UploadStatus.ACTIVE) {
+      throw new UploadNotActiveError();
+    }
+    const data = Buffer.from(dataRaw);
+    this._persistence.appendBytes(upload.path, data);
+    upload.size += data.byteLength;
+    return upload;
+  }
+
+  /**
+   * Queries for an existing resumable upload.
+   * @throws {NotFoundError} if the resumable upload does not exist.
+   */
+  public getResumableUpload(uploadId: string): Upload {
+    return this.findResumableUpload(uploadId);
+  }
+
+  /**
+   * Cancels a resumable upload.
+   * @throws {NotFoundError} if the resumable upload does not exist.
+   * @throws {NotCancellableError} if the resumable upload can not be cancelled.
+   */
+  public cancelResumableUpload(uploadId: string): Upload {
+    const upload = this.findResumableUpload(uploadId);
+    if (upload.status === UploadStatus.FINISHED) {
+      throw new NotCancellableError();
+    }
+    upload.status = UploadStatus.CANCELLED;
+    return upload;
+  }
+
+  /**
+   * Marks a ResumableUpload as finalized.
+   * @throws {NotFoundError} if the resumable upload does not exist.
+   * @throws {NotActiveUploadError} if the resumable upload is not ACTIVE.
+   */
+  public finalizeResumableUpload(uploadId: string): Upload {
+    const upload = this.findResumableUpload(uploadId);
+    if (upload.status === UploadStatus.CANCELLED) {
+      throw new UploadNotActiveError();
+    }
+    upload.status = UploadStatus.FINISHED;
+    return upload;
+  }
+
+  private findResumableUpload(uploadId: string): Upload {
```

Comment:
Not sure why we need this, why not just have `getResumableUpload`?

Reply:
done